File: cocos2d/plugin/plugins/adChina/proj.ios/AdsAdChina.m
/****************************************************************************
 Copyright (c) 2013 cocos2d-x.org
 
 http://www.cocos2d-x.org
 
 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:
 
 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 ****************************************************************************/

#import "AdsAdChina.h"
#import "AdsWrapper.h"
#import "AdManager.h"

#define OUTPUT_LOG(...)     if (self.debug) NSLog(__VA_ARGS__);

@implementation AdsAdChina

@synthesize debug = __debug;
@synthesize popupReady = __popupReady;

- (void) dealloc
{
    if (self.bannerView != nil) {
        [self.bannerView removeFromSuperview];
        [self.bannerView release];
        self.bannerView = nil;
    }

    if (_popupListener != nil)
    {
        [_popupListener release];
    }
    
    if (_fullScreenListener != nil)
    {
        [_fullScreenListener release];
    }
    
    if (nil != self.fullView) {
        [self.fullView removeFromSuperview];
        [self.fullView release];
        self.fullView = nil;
    }

    if (nil != self.popupView) {
        [self.popupView removeFromSuperview];
        [self.popupView release];
        self.popupView = nil;
    }

    [_bannerID release];
    [_fullScreenID release];
    [_popupID release];
    [_offerwallID release];
    [super dealloc];
}

#pragma mark InterfaceAds impl

- (void) configDeveloperInfo: (NSMutableDictionary*) devInfo
{
    self.bannerID = (NSString*) [devInfo objectForKey:@"AdChinaBannerID"];
    self.fullScreenID = (NSString*) [devInfo objectForKey:@"AdChinaFullScreenID"];
    self.popupID = (NSString*) [devInfo objectForKey:@"AdChinaPopupID"];
    //self.offerwallID = (NSString*) [devInfo objectForKey:@"AdChinaOfferwallID"];
        
    [AdManager setExpandLandingPageToolBar:YES];
    self.popupReady= false;
    
    _fullScreenListener = [[AdChinaFullScreenListener alloc] initWithAdsType: @"AdsTypeFullScreen" adsObject:self];
    _popupListener = [[AdChinaInterstitialListener alloc] initWithAdsType: @"AdsTypePopup" adsObject:self];
}

- (void) preloadAds: (NSMutableDictionary*) info
{
    NSNumber* adsType = (NSNumber*) [info objectForKey:@"Param1"];
    NSNumber* adsSize = (NSNumber*) [info objectForKey:@"Param2"];
    NSNumber* adsPos = (NSNumber*) [info objectForKey:@"Param3"];
    
    int numAdsType = [adsType intValue];
    
    switch (numAdsType) {
        case KTypePopup:
            [self preloadPopup];
            break;
        default:
            OUTPUT_LOG(@"The value of 'AdsType' is wrong");
            break;
    }

}

- (void) showAds: (NSMutableDictionary*) info position:(int) pos
{
    NSString* strType = [info objectForKey:@"AdsType"];
    int type = [strType intValue];
    switch (type) {
        case kTypeBanner:
        {
            NSString* strSize = [info objectForKey:@"AdsSizeEnum"];
            int sizeEnum = [strSize intValue];
            [self showBanner:sizeEnum atPos:pos];
            break;
        }
        case kTypeFullScreen:
            [self showFullScreen];
            break;
        case KTypePopup:
            [self showPopup];
            break;
        case KTypeOfferWall:
            [self showOfferWall];
            break;
        default:
            OUTPUT_LOG(@"The value of 'AdsType' is wrong");
            break;
    }
}

- (void) hideAds: (NSMutableDictionary*) info
{
    NSString* strType = [info objectForKey:@"AdsType"];
    int type = [strType intValue];
    switch (type) {
    case kTypeBanner:
        {
            if (nil != self.bannerView) {
                [self.bannerView removeFromSuperview];
                [self.bannerView release];                
                self.bannerView = nil;
                [AdsWrapper onAdsResult:self withRet:kAdsDismissed withMsg:@"AdsTypeBanner"];
            }
            break;
        }
    case kTypeFullScreen:
        {
            if (nil != self.fullView) {
                [self.fullView removeFromSuperview];
                [self.fullView release];
                self.fullView = nil;
            }
            break;
        }
    case KTypePopup:
        {
            if (nil != self.popupView) {
                [self.popupView removeFromSuperview];
                [self.popupView release];
                self.popupView = nil;
            }
            break;
        }
    default:
        OUTPUT_LOG(@"The value of 'AdmobType' is wrong");
        break;
    }
}

- (void) queryPoints
{
    OUTPUT_LOG(@"AdAdChina not support query points!");
}

- (void) spendPoints: (int) points
{
    OUTPUT_LOG(@"points = %lu", (unsigned long)points);
    
    [AdsWrapper onAdsResult:self withRet:kPointsSpendSucceed withMsg:@"spend ponts success!"];
}

- (void) setDebugMode: (BOOL) isDebugMode
{
    self.debug = isDebugMode;
}

- (NSString*) getSDKVersion
{
    return @"3.2.2";
}

- (NSString*) getPluginVersion
{
    return @"0.2.0";
}

- (void) showBanner: (int) sizeEnum atPos:(int) pos
{
    if (nil != self.bannerView) {
        [self.bannerView removeFromSuperview];
        [self.bannerView release];
        self.bannerView = nil;
    }
    
    UIViewController* viewController = [AdsWrapper getCurrentRootViewController];
    CGSize adSize = BannerSizeDefault;
    _bannerPos = pos;
    self.bannerView = [AdChinaBannerView requestAdWithAdSpaceId:_bannerID delegate:self adSize:adSize];
    [self.bannerView retain];
    [self.bannerView setDebugMode:self.debug];
    [self.bannerView setViewControllerForBrowser:viewController];
}

- (void) showFullScreen
{
    if (nil != self.fullView) {
        [self.fullView removeFromSuperview];
        self.fullView = nil;
    }
    
    self.fullView = [AdChinaFullScreenView requestAdWithAdSpaceId:_fullScreenID delegate:_fullScreenListener];
    [self.fullView retain];
    [self.fullView setDebugMode:self.debug];
    [self.fullView setViewControllerForBrowser:[AdsWrapper getCurrentRootViewController]];
}

- (void) showPopup
{
    if (self.popupReady && self.popupView != nil)
    {
        [[AdsWrapper getCurrentRootViewController].view addSubview:self.popupView];
    }
}

- (void) showOfferWall
{
    OUTPUT_LOG(@"AdAdChina not support showOfferWall!");
}

- (void) preloadPopup
{
    if (nil != self.popupView) {
        [self.popupView removeFromSuperview];
        self.popupView = nil;
    }
    
    self.popupView = [AdChinaInterstitialView requestAdWithAdSpaceId:_popupID delegate:_popupListener];
    [self.popupView retain];
    [self.popupView setViewControllerForBrowser:[AdsWrapper getCurrentRootViewController]];
}

#pragma mark AdChinaBanner Delegate impl
- (void)didClickBanner:(AdChinaBannerView *)adView {
    OUTPUT_LOG(@"===AdChina: didClickBanner===");
    [AdsWrapper onAdsResult:self withRet:kAdsClicked withMsg:@"AdsTypeBanner"];
}
     
- (void)didReceiveAd:(AdChinaBannerView *)adView {
    OUTPUT_LOG(@"===AdChina: didGetBanner Tag:%ld===", (long)adView.tag);
    if (self.bannerView != nil)
    {
        [AdsWrapper addAdView:self.bannerView atPos:_bannerPos];
    }
    [AdsWrapper onAdsResult:self withRet:kAdsReceived withMsg:@"AdsTypeBanner"];
}

- (void)didFailedToReceiveAd:(AdChinaBannerView *)adView {
    OUTPUT_LOG(@"===AdChina: didFailToGetBanner Tag:%ld===", (long)adView.tag);
    [AdsWrapper onAdsResult:self withRet:kUnknownError withMsg:@"AdsTypeBanner"];
}
- (void)didEnterFullScreenMode {
    OUTPUT_LOG(@"===AdChina: didEnterFullScreenMode ===");
    [AdsWrapper onAdsResult:self withRet:kAdsEntenLandPage withMsg:@"AdsTypeBanner"];
}
- (void)didExitFullScreenMode {
    OUTPUT_LOG(@"===AdChina: didExitFullScreenMode ===");
    [AdsWrapper onAdsResult:self withRet:kAdsExitLandPage withMsg:@"AdsTypeBanner"];
}
@end

#pragma mark AdChinaFullScreen Delegate impl
@implementation AdChinaFullScreenListener

@synthesize adsType =_adsType;
@synthesize adsObject = _adsObject;
@synthesize debug = __debug;

- (id) init
{
    return [self initWithAdsType:@"AdsTypeFullScreen" adsObject:nil];
}

- (id) initWithAdsType: (NSString*) adsType adsObject: (id) adsObject;
{
    if (self = [super init])
    {
        self.adsType = adsType;
        self.adsObject = adsObject;
    }
    
    return self;
}

- (void)didGetFullScreenAd:(AdChinaFullScreenView *)adView {
    OUTPUT_LOG(@"===AdChina: didGetFullScreenAd===");
    [[AdsWrapper getCurrentRootViewController].view addSubview:adView];
    [AdsWrapper onAdsResult:self.adsObject withRet:kAdsReceived withMsg:self.adsType];
}

- (void)didFailToGetFullScreenAd:(AdChinaFullScreenView *)adView {
    OUTPUT_LOG(@"===AdChina: didFailToGetFullScreenAd===");
    [adView removeFromSuperview];
    [AdsWrapper onAdsResult:self.adsObject withRet:kUnknownError withMsg:self.adsType];
}

- (void)didFinishWatchingFullScreenAd:(AdChinaFullScreenView *)adView {
    OUTPUT_LOG(@"===AdChina: didClosedFullScreenAd===");
    [AdsWrapper onAdsResult:self.adsObject withRet:kAdsDismissed withMsg:self.adsType];
    
}

-(void)didClickFullScreenAd:(AdChinaFullScreenView *)adView {
    OUTPUT_LOG(@"===AdChina: didClickFullScreenAdL===");
    [AdsWrapper onAdsResult:self.adsObject withRet:kAdsClicked withMsg:self.adsType];
}

- (void)didEnterFullScreenMode {
    OUTPUT_LOG(@"=====didEnterFullScreenMode=====");
    [AdsWrapper onAdsResult:self.adsObject withRet:kAdsEntenLandPage withMsg:self.adsType];
}

- (void)didExitFullScreenMode {
    OUTPUT_LOG(@"=====didExitFullScreenMode=====");
    [AdsWrapper onAdsResult:self.adsObject withRet:kAdsExitLandPage withMsg:self.adsType];
}
@end

#pragma mark AdChinaInterstitial Delegate impl

@implementation AdChinaInterstitialListener

@synthesize adsType =_adsType;
@synthesize adsObject = _adsObject;
@synthesize debug = __debug;

- (id) init
{
    return [self initWithAdsType:@"AdsTypePopup" adsObject:nil];
}

- (id) initWithAdsType: (NSString*) adsType adsObject: (id) adsObject;
{
    if (self = [super init])
    {
        self.adsType = adsType;
        self.adsObject = adsObject;
    }
    
    return self;
}

-(void)didGetInterstitialAd:(AdChinaInterstitialView *)adView {
    OUTPUT_LOG(@"=====didGetInterstitialAd======");
    AdsAdChina* ads = (AdsAdChina*)self.adsObject;
    ads.popupReady = true;
    [AdsWrapper onAdsResult:self.adsObject withRet:kAdsReceived withMsg:self.adsType];
}

- (void)didFailToGetInterstitialAd:(AdChinaInterstitialView *)adView {
    OUTPUT_LOG(@"=====didFailToGetInterstitialAd=====");
    [adView removeFromSuperview];
    [AdsWrapper onAdsResult:self.adsObject withRet:kUnknownError withMsg:self.adsType];
}
- (void)didCloseInterstitialAd:(AdChinaInterstitialView *)adView {
    OUTPUT_LOG(@"=====didCloseInterstitialAd=====");
    [adView removeFromSuperview];
    [AdsWrapper onAdsResult:self.adsObject withRet:kAdsDismissed withMsg:self.adsType];
}

// You may use these methods to count click/watch number by yourself
- (void)didClickInterstitialView:(AdChinaInterstitialView *)adView {
    OUTPUT_LOG(@"=====didClickInterstitialView=====");
    [AdsWrapper onAdsResult:self.adsObject withRet:kAdsClicked withMsg:self.adsType];
}

- (void)didEnterFullScreenMode {
    OUTPUT_LOG(@"=====didEnterFullScreenMode=====");
    [AdsWrapper onAdsResult:self.adsObject withRet:kAdsEntenLandPage withMsg:self.adsType];
}

- (void)didExitFullScreenMode {
    OUTPUT_LOG(@"=====didExitFullScreenMode=====");
    [AdsWrapper onAdsResult:self.adsObject withRet:kAdsExitLandPage withMsg:self.adsType];
}
@end

File: cocos2d/plugin/plugins/adChina/proj.android/src/org/cocos2dx/plugin/AdsAdChina.java
/****************************************************************************
Copyright (c) 2012-2013 cocos2d-x.org

http://www.cocos2d-x.org

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
****************************************************************************/
package org.cocos2dx.plugin;

import java.util.HashSet;
import java.util.Hashtable;
import java.util.Iterator;
import java.util.Set;

import com.adchina.android.ads.AdManager;
import com.adchina.android.ads.api.AdBannerListener;
import com.adchina.android.ads.api.AdFsListener;
import com.adchina.android.ads.api.AdFullScreen;
import com.adchina.android.ads.api.AdInterstitial;
import com.adchina.android.ads.api.AdInterstitialListener;
import com.adchina.android.ads.api.AdView;

import android.app.Activity;
import android.content.Context;
import android.location.LocationManager;
import android.util.Log;
import android.view.Gravity;
import android.view.View;
import android.view.WindowManager;
import android.widget.LinearLayout;

public class AdsAdChina implements InterfaceAds {

	private static final String LOG_TAG = "AdsAdChina";
	private static Activity mContext = null;
	private static boolean bDebug = false;
	private static AdsAdChina mAdapter = null;

	private String mBannerID = "";
	private String mFullScreenID = "";
	private String mPopupID = "";
	private String mOfferwallID = "";
	private static AdView mBannerAdView = null;
	private AdFullScreen mFullScreenAdView = null;
	private AdInterstitial mPopupAdView = null;
	private static WindowManager mWm = null;
	private static final int ADCHINA_TYPE_BANNER = 1;
	private static final int ADCHINA_TYPE_FULLSCREEN = 2;
	private static final int ADCHINA_TYPE_POPUP = 3;
	private static final int ADCHINA_TYPE_OFFERWALL = 4;
    
	protected static void LogE(String msg, Exception e) {
		Log.e(LOG_TAG, msg, e);
		e.printStackTrace();
	}

	protected static void LogD(String msg) {
		if (bDebug) {
			Log.d(LOG_TAG, msg);
		}
	}

	public AdsAdChina(Context context) {
		mContext = (Activity) context;
		mAdapter = this;
	}
	
	@Override
	public void setDebugMode(boolean debug) {
	    LogD("setDebugMode" + debug);
		bDebug = debug;
		AdManager.setLogMode(debug);   
	}

	@Override
	public String getSDKVersion() {
		return "3.2.1";
	}

	@Override
	public void configDeveloperInfo(Hashtable<String, String> devInfo) {
		try {
		    mBannerID = devInfo.get("AdChinaBannerID");
		    mFullScreenID = devInfo.get("AdChinaFullScreenID");
		    mPopupID = devInfo.get("AdChinaPopupID");
		    mOfferwallID = devInfo.get("AdChinaOfferwallID");
		    AdManager.setCanHardWare(true);
		    AdManager.setExpandToolBar(true);
		    AdManager.setAnimation(true);
		    AdManager.setEnableLbs(true);
		    AdManager.setLocationManager((LocationManager)mContext.getSystemService(Context.LOCATION_SERVICE));
		} catch (Exception e) {
			LogE("initAppInfo, The format of appInfo is wrong", e);
		}
	}

	@Override
	public void showAds(Hashtable<String, String> info, int pos) {
	    if (! NetworkUtility.networkReachable(mContext)) 
        {
            return;
        }
	    try
	    {
	        String strType = info.get("AdsType");
	        int adsType = Integer.parseInt(strType);
	            
	        switch (adsType) {
    	        case ADCHINA_TYPE_BANNER:
    	            {
    	                String strSize = info.get("AdsSizeEnum");
                        int sizeEnum = Integer.parseInt(strSize);
    	                showBannerAd(sizeEnum, pos);
                        break;
    	            }
    	        case ADCHINA_TYPE_FULLSCREEN:
    	            showFullScreenAd();
    	            break;
    	        case ADCHINA_TYPE_POPUP:
    	            showPopUpAd();
                    break;
    	        case ADCHINA_TYPE_OFFERWALL:
    	            LogD("AdChina not support showOfferwall!");
    	            break;
    	        default:
    	            break;
	        }
	    } catch (Exception e) {
	        LogE("Error when show Ads ( " + info.toString() + " )", e);
	    }
	}

	@Override
	public void spendPoints(int points) {
		LogD("AdChina not support spend points!");
	}

    @Override
    public void queryPoints() {
        LogD("AdChina not support query points!");
    }
    
	@Override
	public void hideAds(Hashtable<String, String> info) {
	    try
        {
            String strType = info.get("AdsType");
            int adsType = Integer.parseInt(strType);

            switch (adsType) {
            case ADCHINA_TYPE_BANNER:
                hideBannerAd();
                break;
            default:
                break;
            }
        } catch (Exception e) {
            LogE("Error when hide Ads ( " + info.toString() + " )", e);
        }
	}
    
	public void requestAd(int adsType)
    {
        try
        {
            switch (adsType) {
                case ADCHINA_TYPE_BANNER:
                    {
                      
                        break;
                    }
                case ADCHINA_TYPE_FULLSCREEN:
                    
                    break;
                case ADCHINA_TYPE_POPUP:
                    
                    break;
                case ADCHINA_TYPE_OFFERWALL:
                    
                    break;
                default:
                    break;
                
            }
        } catch (Exception e) {
            LogE("Error when request Ads ( " + adsType + " )", e);
        }
    }
    
	private void showFullScreenAd()
	{
	    PluginWrapper.runOnMainThread(new Runnable() {

            @Override
            public void run() {
                mFullScreenAdView = new AdFullScreen(mContext, mFullScreenID);
                mFullScreenAdView.setAdFsListener(new AdChinaFsListenerImpl());
                mFullScreenAdView.start();
            }
	    });
	}
	
	private void showPopUpAd()
	{
	    PluginWrapper.runOnMainThread(new Runnable() {

            @Override
            public void run() {
                mPopupAdView = new AdInterstitial(mContext, mPopupID);
                mPopupAdView.setAdInterstitialListener(new AdChinaInterstitialListenerImpl());
                mPopupAdView.start();
            }
        });
	}

	private void showBannerAd(int sizeEnum, int pos) {
		final int curPos = pos;
		final int curSize = sizeEnum;

		PluginWrapper.runOnMainThread(new Runnable() {

            @Override
			public void run() {		    
                
                if (null != mBannerAdView) {
                    if (null != mWm) {
                        mWm.removeView(mBannerAdView);
                    }
                    mBannerAdView = null;
                }
                
                if (null == mWm) {
                    mWm = (WindowManager) mContext.getSystemService("window");
                }               
                
				mBannerAdView = new AdView(mContext, mBannerID, false, false);
				AdsWrapper.addAdView(mWm, mBannerAdView, curPos);
				
				mBannerAdView.setAdBannerListener(new AdChinaBannerListenerImpl());
				mBannerAdView.start();
				mBannerAdView.setVisibility(View.GONE);
			}
		});
	}

	private void hideBannerAd() {
	    PluginWrapper.runOnMainThread(new Runnable() {
            @Override
            public void run() {
                if (null != mBannerAdView) {
                    if (null != mWm) {
                        mWm.removeView(mBannerAdView);
                    }
                    mBannerAdView.stop();
                    mBannerAdView = null;
                }
            }
        });
	}

	
	private class AdChinaBannerListenerImpl implements AdBannerListener {

		@Override
		public void onClickBanner(AdView arg0) {
			LogD("onClickBanner invoked");
			AdsWrapper.onAdsResult(mAdapter, AdsWrapper.RESULT_CODE_AdsClicked, "AdsTypeBanner");
		}

		@Override
		public void onFailedToReceiveAd(AdView arg0) {
			LogD("onFailedToReceiveAd invoked");
			AdsWrapper.onAdsResult(mAdapter, AdsWrapper.RESULT_CODE_UnknownError, "AdsTypeBanner");
		}

		@Override
		public void onReceiveAd(AdView arg0) {
			LogD("onReceiveAd invoked");
			if (arg0 != null)
			{
			    arg0.setVisibility(View.VISIBLE);
			}
			AdsWrapper.onAdsResult(mAdapter, AdsWrapper.RESULT_CODE_AdsReceived, "AdsTypeBanner");
		}
	}

	private class AdChinaInterstitialListenerImpl implements AdInterstitialListener {

        @Override
        public void onClickItst() {
            LogD("onClickItst invoked");
            AdsWrapper.onAdsResult(mAdapter, AdsWrapper.RESULT_CODE_AdsClicked, "AdsTypePopup");
        }

        @Override
        public void onFailedToReceiveItstAd() {
            LogD("onFailedToReceiveItstAd invoked");
            AdsWrapper.onAdsResult(mAdapter, AdsWrapper.RESULT_CODE_UnknownError, "AdsTypePopup");
        }

        @Override
        public void onReceivedItstAd() {
            LogD("onReceivedItstAd invoked");
            if(mPopupAdView != null)
            {
                mPopupAdView.showItst();
            }
            AdsWrapper.onAdsResult(mAdapter, AdsWrapper.RESULT_CODE_AdsReceived, "AdsTypePopup");
        }
        
        @Override
        public void onCloseItst() {
            LogD("onCloseItst invoked");
            if(mPopupAdView != null)
            {
                mPopupAdView.stop();
            }
            AdsWrapper.onAdsResult(mAdapter, AdsWrapper.RESULT_CODE_AdsDismissed, "AdsTypePopup");
        }

        @Override
        public void onDisplayItst() {
            LogD("onDisplayItst invoked");
            AdsWrapper.onAdsResult(mAdapter, AdsWrapper.RESULT_CODE_AdsShown, "AdsTypePopup");
        }
    }
	
	private class AdChinaFsListenerImpl implements AdFsListener {

        @Override
        public void onClickFullScreenAd() {
            LogD("onClickFullScreenAd invoked");
            AdsWrapper.onAdsResult(mAdapter, AdsWrapper.RESULT_CODE_AdsClicked, "AdsTypeFullScreen");
        }

        @Override
        public void onFailedToReceiveFullScreenAd() {
            LogD("onFailedToReceiveFullScreenAd invoked");
            AdsWrapper.onAdsResult(mAdapter, AdsWrapper.RESULT_CODE_UnknownError, "AdsTypeFullScreen");
        }

        @Override
        public void onEndFullScreenLandpage() {
            LogD("onEndFullScreenLandpage invoked");  
        }
        
        @Override
        public void onStartFullScreenLandPage() {
            LogD("onStartFullScreenLandPage invoked");     
        }
        
        @Override
        public void onReceiveFullScreenAd() {
            LogD("onReceiveFullScreenAd invoked");
            if(mFullScreenAdView != null)
            {
                mFullScreenAdView.showFs();
            }
            AdsWrapper.onAdsResult(mAdapter, AdsWrapper.RESULT_CODE_AdsReceived, "AdsTypeFullScreen!");
        }

        @Override
        public void onDisplayFullScreenAd() {
            LogD("onDisplayFullScreenAd invoked");
            AdsWrapper.onAdsResult(mAdapter, AdsWrapper.RESULT_CODE_AdsShown, "AdsTypeFullScreen");
        }

        @Override
        public void onFinishFullScreenAd() {
            LogD("onFinishFullScreenAd invoked");
            if(mFullScreenAdView != null)
            {
                mFullScreenAdView.stop();
            }
            AdsWrapper.onAdsResult(mAdapter, AdsWrapper.RESULT_CODE_AdsDismissed, "AdsTypeFullScreen");        
        }
    }
	
	@Override
	public String getPluginVersion() {
		return "0.2.0";
	}
	
	public static void dismiss()
    {
        LogD("onDestory AdChina");
        if (null != mBannerAdView) {
            if (null != mWm) {
                mWm.removeView(mBannerAdView);
            }
            mBannerAdView = null;
        }
    }
}
